cluster_management/src/main/java/com/pinterest/rocksplicator/BulkIngestionStateModelFactory.java

```diff
@@ -0,0 +1,150 @@
+    private static final Logger LOG = LoggerFactory.getLogger(BulkIngestionStateModel.class);
+
+    private final String resourceName;
+    private final String partitionName;
+    private final int adminPort;
+    private final String cluster;
+    private final String host;
+
+    /**
+     * State model that handles the state machine of a single partition.
+     */
+    public BulkIngestionStateModel(String resourceName, String partitionName, int adminPort,
+                                   String cluster, CuratorFramework zkClient, String host) {
+      super(resourceName, partitionName, adminPort, cluster, zkClient);
+      this.resourceName = resourceName;
+      this.partitionName = partitionName;
+      this.host = host;
+      this.adminPort = adminPort;
+      this.cluster = cluster;
+    }
+
+    /**
+     * Callback for OFFLINE to ONLINE transition.
+     * This can happen when 1) a new partition is first launched on a host; or 2) an existing
+     * partition is reopened when restart/zk temporarily expires.
+     * We first make sure the DB is open and then load SST files into the DB.
+     */
+    @Override
+    public void onBecomeOnlineFromOffline(Message message, NotificationContext context) {
+      super.checkSanity("OFFLINE", "ONLINE", message);
+      Utils.logTransitionMessage(message);
+      Utils.addDB(Utils.getDbName(partitionName), adminPort);
+
+      try {
+        String dbName = Utils.getDbName(partitionName);
+
+        // open the DB if it's currently not opened yet
+        Utils.addDB(dbName, adminPort);
+
+        // see if the DB needs to be rebuilt. If there is anything in the meta,
+        // assume that we have a complete local replica. We can ensure this
+        // by locking the helix shard config during data uploads
+        CheckDBResponse localStatus = Utils.checkLocalDB(dbName, adminPort);
+
+        // check that the metadata is set and well formed, otherwise we
+        // build a replica from another host serving.
+        boolean needRebuild = !localStatus.isSetDb_meta_data();
+        needRebuild = needRebuild && !localStatus.getDb_meta_data().isSetS3_bucket();
+        needRebuild = needRebuild && !localStatus.getDb_meta_data().isSetS3_path();
+        if (needRebuild) {
+          HelixAdmin admin = context.getManager().getClusterManagmentTool();
+          ExternalView view = admin.getResourceExternalView(cluster, resourceName);
+          Map<String, String> stateMap = view.getStateMap(partitionName);
+          // find live replicas
+
+          for (Map.Entry<String, String> instanceNameAndRole : stateMap.entrySet()) {
+            String role = instanceNameAndRole.getValue();
+            if (!role.equalsIgnoreCase("ONLINE")) {
+              continue;
+            }
+
+            String hostPort = instanceNameAndRole.getKey();
+            String host = hostPort.split("_")[0];
```

Comment:
perform the split and check the size of the array before using the values.
Also can we check for a null value from .getKey() using Preconditions.checkNotEmpty() or something similiar